gtk/details.cc

```diff
@@ -1455,323 +1281,253 @@ static void refreshPeerRow(GtkListStore* store, GtkTreeIter* iter, tr_peer_stat
-                gtk_tree_path_free(p);
+                auto const iter = store->append();
+                (*iter)[webseed_cols.url] = url;
+                (*iter)[webseed_cols.key] = key;
+                hash.emplace(key, Gtk::TreeRowReference(store, store->get_path(iter)));
             }
         }
     }
 
     /* step 3: update the webseeds */
-    for (int i = 0; i < n; ++i)
+    for (size_t i = 0; i < torrents.size(); ++i)
     {
-        tr_torrent const* tor = torrents[i];
-        tr_info const* inf = tr_torrentInfo(tor);
+        auto const* tor = torrents.at(i);
+        auto const* inf = tr_torrentInfo(tor);
         double* speeds_KBps = tr_torrentWebSpeeds_KBps(tor);
 
         for (unsigned int j = 0; j < inf->webseedCount; ++j)
         {
             char const* const url = inf->webseeds[j];
-
-            char key[256];
-            g_snprintf(key, sizeof(key), "%d.%s", tr_torrentId(tor), url);
-            auto* const ref = static_cast<GtkTreeRowReference*>(g_hash_table_lookup(hash, key));
-            GtkTreePath* const p = gtk_tree_row_reference_get_path(ref);
-            gtk_tree_model_get_iter(model, &iter, p);
+            auto const key = gtr_sprintf("%d.%s", tr_torrentId(tor), url);
+            auto const iter = store->get_iter(hash.at(key).get_path());
 
             char buf[128] = { 0 };
             if (speeds_KBps[j] > 0)
             {
                 tr_formatter_speed_KBps(buf, speeds_KBps[j], sizeof(buf));
             }
 
-            gtk_list_store_set(
-                store,
-                &iter,
-                TR_ARG_TUPLE(WEBSEED_COL_DOWNLOAD_RATE_DOUBLE, speeds_KBps[j]),
-                TR_ARG_TUPLE(WEBSEED_COL_DOWNLOAD_RATE_STRING, buf),
-                TR_ARG_TUPLE(WEBSEED_COL_WAS_UPDATED, TRUE),
-                -1);
-
-            gtk_tree_path_free(p);
+            (*iter)[webseed_cols.download_rate_double] = speeds_KBps[j];
+            (*iter)[webseed_cols.download_rate_string] = buf;
+            (*iter)[webseed_cols.was_updated] = true;
         }
 
         tr_free(speeds_KBps);
     }
 
     /* step 4: remove webseeds that have disappeared */
-    if (gtk_tree_model_iter_nth_child(model, &iter, nullptr, 0))
+    if (auto iter = store->children().begin(); iter)
     {
-        gboolean more = TRUE;
-
-        while (more)
+        while (iter)
         {
-            gboolean b;
-            gtk_tree_model_get(model, &iter, WEBSEED_COL_WAS_UPDATED, &b, -1);
-
-            if (b)
+            if (iter->get_value(webseed_cols.was_updated))
             {
-                more = gtk_tree_model_iter_next(model, &iter);
+                ++iter;
             }
             else
             {
-                char* key;
-                gtk_tree_model_get(model, &iter, WEBSEED_COL_KEY, &key, -1);
+                auto const key = iter->get_value(webseed_cols.key);
 
-                if (key != nullptr)
+                if (!key.empty())
```

Comment:
_nit_ this safety check was necessary iirc because g_hash_table_remove would crash otherwise. We could knock one point off the comlexity ranking by removing the check here I think?